piker/brokers/kucoin.py

```diff
@@ -0,0 +1,685 @@
+
+        log.info(f' {len(syms)} Kucoin market pairs fetched')
+        return syms
+
+    async def cache_pairs(
+        self,
+    ) -> dict[str, KucoinMktPair]:
+        '''
+        Get cached pairs and convert keyed symbols into fqsns if ya want
+
+        '''
+        if not self._pairs:
+            self._pairs = await self._get_pairs()
+
+        return self._pairs
+
+    async def search_symbols(
+        self,
+        pattern: str,
+        limit: int = 30,
+    ) -> dict[str, KucoinMktPair]:
+        data = await self.cache_pairs()
+
+        matches = fuzzy.extractBests(
+            pattern, data, score_cutoff=35, limit=limit)
+        # repack in dict form
+        return {item[0].name: item[0] for item in matches}
+
+    async def last_trades(self, sym: str) -> list[AccountTrade]:
+        trades = await self._request('GET', f'/accounts/ledgers?currency={sym}', 'v1')
+        trades = AccountResponse(**trades)
+        return trades.items
+
+    async def _get_bars(
+        self,
+        fqsn: str,
+        start_dt: datetime | None = None,
+        end_dt: datetime | None = None,
+        limit: int = 1000,
+        as_np: bool = True,
+        type: str = '1min',
+    ) -> np.ndarray:
+        '''
+        Get OHLC data and convert to numpy array for perffff:
+        https://docs.kucoin.com/#get-klines
+
+        Kucoin bar data format:
+        [
+             "1545904980",             //Start time of the candle cycle 0
+             "0.058",                  //opening price 1
+             "0.049",                  //closing price 2
+             "0.058",                  //highest price 3
+             "0.049",                  //lowest price 4
+             "0.018",                  //Transaction volume 5
+             "0.000945"                //Transaction amount 6
+        ],
+
+        piker ohlc numpy array format:
+        [
+            ('index', int),
+            ('time', int),
+            ('open', float),
+            ('high', float),
+            ('low', float),
+            ('close', float),
+            ('volume', float),
+            ('bar_wap', float),  # will be zeroed by sampler if not filled
+        ]
+
+        '''
+        # Generate generic end and start time if values not passed
+        # Currently gives us 12hrs of data
+        if end_dt is None:
+            end_dt = pendulum.now('UTC').add(minutes=1)
+
+        if start_dt is None:
+            start_dt = end_dt.start_of('minute').subtract(minutes=limit)
+
+        start_dt = int(start_dt.timestamp())
+        end_dt = int(end_dt.timestamp())
+
+        kucoin_sym = fqsn_to_kucoin_sym(fqsn, self._pairs)
+
+        url = f'/market/candles?type={type}&symbol={kucoin_sym}&startAt={start_dt}&endAt={end_dt}'
+
+        for i in range(10):
+            data = await self._request(
+                'GET',
+                url,
+                api_v='v1',
+            )
+
+            if not isinstance(data, list):
```

Comment:
what is the normal type of `data` ?
would be handy to type annot this even if it is somewhat variable 😉